Mathlib/Order/BooleanSubalgebra.lean

```diff
@@ -0,0 +1,379 @@
+
+@[simp] lemma map_map (g : BoundedLatticeHom β γ) (f : BoundedLatticeHom α β) :
+    (L.map f).map g = L.map (g.comp f) := SetLike.coe_injective <| image_image _ _ _
+
+lemma mem_map_equiv {f : α ≃o β} {a : β} : a ∈ L.map f ↔ f.symm a ∈ L := Set.mem_image_equiv
+
+lemma apply_mem_map_iff (hf : Injective f) : f a ∈ L.map f ↔ a ∈ L := hf.mem_set_image
+
+lemma map_equiv_eq_comap_symm (f : α ≃o β) (L : BooleanSubalgebra α) :
+    L.map f = L.comap (f.symm : BoundedLatticeHom β α) :=
+  SetLike.coe_injective <| f.toEquiv.image_eq_preimage L
+
+lemma comap_equiv_eq_map_symm (f : β ≃o α) (L : BooleanSubalgebra α) :
+    L.comap f = L.map (f.symm : BoundedLatticeHom α β) := (map_equiv_eq_comap_symm f.symm L).symm
+
+lemma map_symm_eq_iff_eq_map {M : BooleanSubalgebra β} {e : β ≃o α} :
+    L.map ↑e.symm = M ↔ L = M.map ↑e := by
+  simp_rw [← coe_inj]; exact (Equiv.eq_image_iff_symm_image_eq _ _ _).symm
+
+lemma map_le_iff_le_comap {f : BoundedLatticeHom α β} {M : BooleanSubalgebra β} :
+    L.map f ≤ M ↔ L ≤ M.comap f := image_subset_iff
+
+lemma gc_map_comap (f : BoundedLatticeHom α β) : GaloisConnection (map f) (comap f) :=
+  fun _ _ ↦ map_le_iff_le_comap
+
+@[simp] lemma map_bot (f : BoundedLatticeHom α β) : (⊥ : BooleanSubalgebra α).map f = ⊥ :=
+  (gc_map_comap f).l_bot
+
+lemma map_sup (f : BoundedLatticeHom α β) (L M : BooleanSubalgebra α) :
+    (L ⊔ M).map f = L.map f ⊔ M.map f := (gc_map_comap f).l_sup
+
+lemma map_iSup (f : BoundedLatticeHom α β) (L : ι → BooleanSubalgebra α) :
+    (⨆ i, L i).map f = ⨆ i, (L i).map f := (gc_map_comap f).l_iSup
+
+@[simp] lemma comap_top (f : BoundedLatticeHom α β) : (⊤ : BooleanSubalgebra β).comap f = ⊤ :=
+  (gc_map_comap f).u_top
+
+lemma comap_inf (L M : BooleanSubalgebra β) (f : BoundedLatticeHom α β) :
+    (L ⊓ M).comap f = L.comap f ⊓ M.comap f := (gc_map_comap f).u_inf
+
+lemma comap_iInf (f : BoundedLatticeHom α β) (L : ι → BooleanSubalgebra β) :
+    (⨅ i, L i).comap f = ⨅ i, (L i).comap f := (gc_map_comap f).u_iInf
+
+lemma map_inf_le (L M : BooleanSubalgebra α) (f : BoundedLatticeHom α β) :
+    map f (L ⊓ M) ≤ map f L ⊓ map f M := map_mono.map_inf_le _ _
+
+lemma le_comap_sup (L M : BooleanSubalgebra β) (f : BoundedLatticeHom α β) :
+    comap f L ⊔ comap f M ≤ comap f (L ⊔ M) := comap_mono.le_map_sup _ _
+
+lemma le_comap_iSup (f : BoundedLatticeHom α β) (L : ι → BooleanSubalgebra β) :
+    ⨆ i, (L i).comap f ≤ (⨆ i, L i).comap f := comap_mono.le_map_iSup
+
+lemma map_inf (L M : BooleanSubalgebra α) (f : BoundedLatticeHom α β) (hf : Injective f) :
+    map f (L ⊓ M) = map f L ⊓ map f M := by
+  rw [← SetLike.coe_set_eq]
+  simp [Set.image_inter hf]
+
+lemma map_top (f : BoundedLatticeHom α β) (h : Surjective f) : BooleanSubalgebra.map f ⊤ = ⊤ :=
+  SetLike.coe_injective <| by simp [h.range_eq]
+
+/-- The minimum boolean algebra containing a given set. -/
+def closure (s : Set α) : BooleanSubalgebra α := sInf {L | s ⊆ L}
```

Comment:
Doesn't the closure operator stuff help here?

Reply:
No, because we're not returning a `Set` but a `BooleanSubalgebra`. If you remember, the return type is the distinction I draw between Galois insertions and closure operators